File: csharp/AoC2021/Day17/Day17/TargetArea.cs
namespace Day17;

public class TargetArea
{
    private readonly long _xMin;
    private readonly long _xMax;
    private readonly long _yMin;
    private readonly long _yMax;

    public long FindMaxHeight()
    {
        var (d, y) = GetYBound();
        return y switch
        {
            < 0 => d * (d - 1) / 2,
            > 0 => d * (d + 1) / 2,
            _ => throw new InvalidOperationException(
                "The target area contains a point with y=0. This means the trajectory can be arbitrarily high."),
        };
    }

    public IEnumerable<(long, long)> FindAllValidTrajectories()
    {
        var trajectories = new List<(long, long)>();
        var (d, _) = GetYBound();

        for (var x = 0; x <= _xMax; x += 1)
        {
            for (var y = -d; y < d; y += 1)
            {
                if (IsValidTrajectory((x, y)))
                    trajectories.Add((x, y));
            }
        }

        return trajectories;
    }

    private (long Distance, long Y) GetYBound()
    {
        if (_yMin < 0 && _yMax < 0)
            return (Math.Abs(_yMin), _yMin);

        if (_yMin > 0 && _yMax > 0)
            return (Math.Abs(_yMax), _yMax);

        return (0, 0);
    }

    private bool IsValidTrajectory((long VX, long VY) initial)
    {
        var (vX, vY) = initial;
        var x = 0L;
        var y = 0L;

        while (true)
        {
            x += vX;
            y += vY;

            if (x > _xMax) // we have overshot the target area
                return false;

            if (vY < 0 && y < _yMin) // we have overshot the target area
                return false;

            if (x >= _xMin && y >= _yMin && y <= _yMax)
                return true;

            if (vX == 0 && x < _xMin) // x will no longer change, so we will never reach the target area
                return false;

            if (vX != 0)
                vX -= vX / Math.Abs(vX);
            vY -= 1;
        }
    }

    private TargetArea(long xMin, long xMax, long yMin, long yMax)
    {
        _xMin = xMin;
        _xMax = xMax;
        _yMin = yMin;
        _yMax = yMax;
    }

    public static TargetArea Parse(IEnumerable<string> lines)
    {
        var line = lines.First();

        // find relevant indices (correct input is assumed, so no -1 checks are done)
        var i0 = 2 + line.IndexOf("x=", StringComparison.Ordinal);
        var i1 = line.IndexOf("..", i0, StringComparison.Ordinal);
        var i2 = i1 + 2;
        var i3 = line.IndexOf(", y=", i2, StringComparison.Ordinal);
        var i4 = i3 + 4;
        var i5 = line.IndexOf("..", i4, StringComparison.Ordinal);
        var i6 = i5 + 2;

        // parse values
        var xMin = long.Parse(line[i0..i1]);
        var xMax = long.Parse(line[i2..i3]);
        var yMin = long.Parse(line[i4..i5]);
        var yMax = long.Parse(line[i6..]);

        return new TargetArea(xMin, xMax, yMin, yMax);
    }
}

File: rust/aoc2021/src/day17/target_area.rs
use std::cmp::Ordering;
use std::io;

use itertools::Itertools;
use rdcl_aoc_helpers::error::ParseError;

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct TargetArea {
    x_min: i64,
    x_max: i64,
    y_min: i64,
    y_max: i64,
}

impl TargetArea {
    /// To find the maximum height, we just need to focus on the vertical velocity. The horizontal
    /// velocity does not impact the height of the trajectory. In principle we should verify that
    /// the trajectory does go through the target area, considering the horizontal component, but we
    /// can probably ignore this for this puzzle.
    ///
    /// For the vertical velocity, we know that once the probe reaches its maximum height, its
    /// vertical velocity will be 0. So we don't actually care about its initial velocity, so long
    /// as the maximum height can actually be reached from the starting position, and the probe
    /// falls through the target area on its way down.
    ///
    /// When falling down, the probe will follow the n(n+1)/2 formula (every step the height
    /// decreases with one more than the previous step). Since the probe has to pass through 0 and
    /// the target area, we should find the point in the target area that is furthest from 0. The
    /// distance between this point and 0 gives us n.
    pub fn find_max_height(&self) -> i64 {
        let (d, y) = self.get_y_bound();
        match y.cmp(&0) {
            Ordering::Less => d * (d - 1) / 2,
            Ordering::Equal => panic!("The target area contains a point with y=0. This means the trajectory can be arbitrarily high."),
            Ordering::Greater => d * (d + 1) / 2,
        }
    }

    /// In order to find all valid initial velocities, we need to have some bounds on the search
    /// area. For the initial horizontal velocity, this is easy: The initial velocity must be low
    /// enough to not overshoot the target area in the first step. For the vertical velocity we can
    /// use the information from the previous part (the velocity that reaches the max height is also
    /// the max velocity).
    ///
    /// TODO: Optimize. (E.g. pick a better lower bound for v_x.)
    pub fn find_all_valid_initial_velocities(&self) -> Vec<(i64, i64)> {
        let (d, _) = self.get_y_bound();
        (0..=self.x_max)
            .cartesian_product(-d..d)
            .filter(|&c| self.is_valid_trajectory(c))
            .collect()
    }

    fn get_y_bound(&self) -> (i64, i64) {
        if self.y_min < 0 && self.y_max < 0 {
            (self.y_min.abs(), self.y_min)
        } else if self.y_min > 0 && self.y_max > 0 {
            (self.y_max.abs(), self.y_max)
        } else {
            (0, 0)
        }
    }

    fn is_valid_trajectory(&self, v_initial: (i64, i64)) -> bool {
        let (mut v_x, mut v_y) = v_initial;
        let mut x = 0;
        let mut y = 0;

        loop {
            x += v_x;
            y += v_y;

            if x > self.x_max {
                // we have overshot the target area
                return false;
            }

            if v_y < 0 && y < self.y_min {
                // we have overshot the target area
                return false;
            }

            if x >= self.x_min && y >= self.y_min && y <= self.y_max {
                return true;
            }

            if v_x == 0 && x < self.x_min {
                // x will no longer change, so we will never reach the target area
                return false;
            }

            if v_x != 0 {
                v_x -= v_x / v_x.abs();
            }
            v_y -= 1;
        }
    }

    pub fn parse<I>(mut lines: I) -> Result<TargetArea, ParseError>
    where
        I: Iterator<Item = io::Result<String>>,
    {
        let line = lines.next().unwrap()?;
        let s = line.strip_prefix("target area: x=").unwrap();
        let p1 = s.find("..").unwrap();
        let x_min = s[0..p1].parse()?;
        let p2 = s.find(", y=").unwrap();
        let x_max = s[p1 + 2..p2].parse()?;
        let p3 = p2 + 4;
        let p4 = p3 + s[p3..].find("..").unwrap();
        let y_min = s[p3..p4].parse()?;
        let y_max = s[p4 + 2..].parse()?;

        Ok(TargetArea {
            x_min,
            x_max,
            y_min,
            y_max,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_find_initial_velocity_with_max_height() {
        assert_eq!(target_area().find_max_height(), 45);
    }

    #[test]
    fn test_find_all_valid_initial_velocities() {
        assert_eq!(target_area().find_all_valid_initial_velocities().len(), 112);
    }

    #[test]
    fn test_parse_input() {
        let input = vec![Ok("target area: x=20..30, y=-10..-5".to_string())];
        assert_eq!(TargetArea::parse(input.into_iter()), Ok(target_area()));
    }

    fn target_area() -> TargetArea {
        TargetArea {
            x_min: 20,
            x_max: 30,
            y_min: -10,
            y_max: -5,
        }
    }
}
